source/server/session_resource_repository.h

```diff
@@ -0,0 +1,221 @@
+#pragma once
+
+#include <mutex>
+
+#include "session_repository.h"
+
+namespace nidevice_grpc {
+
+// Wraps SessionRepository to store a service-specific TResourceHandle associated with
+// each session.
+template <typename TResourceHandle>
+class SessionResourceRepository {
+ public:
+  typedef std::function<std::tuple<int, TResourceHandle>()> InitFunc;
+  typedef std::function<void(TResourceHandle)> CleanupSessionFunc;
+
+  SessionResourceRepository(SessionRepository* session_repository)
+      : session_repository_(session_repository)
+  {
+  }
+
+  int add_session(
+      const std::string& session_name,
+      InitFunc init_func,
+      CleanupSessionFunc cleanup_func,
+      uint32_t& session_id);
+
+  TResourceHandle access_session(uint32_t session_id, const std::string& session_name) const;
+  uint32_t resolve_session_id(TResourceHandle handle) const;
+
+  void remove_session(TResourceHandle handle);
+
+ private:
+  // Wraps init_func_ and caches the result as handle_.
+  // For passing into SessionRepository.add_session.
+  struct SessionResourceCreator {
+    const InitFunc& init_func_;
+    TResourceHandle handle_;
+    bool added_new_handle_;
+
+    SessionResourceCreator(const InitFunc& init_func)
+        : init_func_(init_func),
+          added_new_handle_(false),
+          handle_()
+    {
+    }
+
+    uint32_t init_resource_handle();
+  };
+
+  // A thread-safe bi-directional map of session_id to TResourceHandle.
+  class ResourceMap {
+   public:
+    void add(uint32_t session_id, TResourceHandle handle);
+
+    bool contains(uint32_t session_id) const;
+    TResourceHandle get_handle(uint32_t session_id) const;
+    uint32_t get_session_id(TResourceHandle handle) const;
+
+    uint32_t remove_handle(TResourceHandle handle);
+    TResourceHandle remove_session_id(uint32_t session_id);
+
+   private:
+    using SessionToResourceMap = std::map<uint32_t, TResourceHandle>;
+    using ResourceToSessionMap = std::map<TResourceHandle, uint32_t>;
+    using MapLock = std::lock_guard<std::recursive_mutex>;
+
+    SessionToResourceMap map_;
+    ResourceToSessionMap reverse_map_;
+    mutable std::recursive_mutex map_mutex_;
+  };
+
+  SessionRepository* session_repository_;
+  ResourceMap resource_map_;
+};
+
+template <typename TResourceHandle>
```

Comment:
Note: [Here's ](https://google.github.io/styleguide/cppguide.html#Self_contained_Headers) the style guide section that covers this.